lib/pbench/agent/tool_meister_start.py

```diff
@@ -296,8 +295,8 @@ def start_tms_via_ssh(
     cmd = f"{tool_meister_cmd} {redis_server.host} {redis_server.port} {{tm_param_key}} yes"
     if debug_level:
         cmd += f" {debug_level}"
-    template = TemplateSsh(ssh_cmd, shlex.split(ssh_opts), cmd)
-    tms = dict()
+    template = TemplateSsh(Path(ssh_cmd), shlex.split(ssh_opts), cmd)
+    tms: Dict[str, Union[str, int, Dict[str, str]]] = {}
```

Comment:
Does this type hint really help?  This is so hard to parse that it's value is dubious.

Reply:
I agree some of these are getting a bit gnarly for human readers, but I think part of this is motivated by the fact that our tools are getting smarter; I (and I believe Webb) use IDEs that have started to diagnose potential problems with the use of variables due to our lax type hints. (E.g., this may be a type that doesn't have `.get` or can't be indexed or whatever.) The result is that we're seeing a lot of red representing false problems just because we haven't properly informed the IDE of our intent.

I think our code base is still a ways from being able to put a strong type-hint aware lint checker into our Jenkins pipeline and have any chance of success; however I'd love to be able to work towards doing that. "Picking lint" like this along the way moves us towards better (testable) self-documentation, and I look at that as a good thing.

We could probably help to make it more readable with well-chosen typedefs as Webb and I tried to do with the `JSON` family. (Unfortunately this isn't *quite* a `JSONOBJECT` since we're tying this one to `Dict[str, str]`.)

Reply:
I thought the whole point of a language was to make it so that a human being can write in it reasonably, and the computer does the hard work of transforming what is written into what it can execute.

That is one reason why I am desperately trying to drop Perl!

These Python3 Type Hints seem to be taking Python the other direction: you need a IDE to derive value for what is written?

I am just not buying it.  I'll resolve this issue, but it is just not the right direction for the language.

Reply:
A language should be readable, and it should be semantically robust enough to support technologies that help prevent or diagnose coding errors at build time rather than at run time. It's why we run `flake8` and `LGTM` ... although the degree to which they can help us is significantly limited by Python's flexibility in typing and name binding.

Perl and Python are both originally "scripting" languages, not "software development" languages: they were designed to be easy and quick to write and hack, without much concentration on supporting readability, maintainability, or important technologies like static analysis. The fact that a name can bind to anything is a great convenience for throwaway code; it's a massive complication for long-term maintenance and growth. Type hints allow unambiguous documentation both for the reader and for tools of what the developer *intended*, instead of requiring a reader to study the internal code and call points to figure out exactly how to call a function.

The fact that Python itself doesn't care can be characterized as a throwback to the "throwaway scripting language" mentality, although, to be fair, there's also a substantial backwards compatibility issue.

Python 3.9 will make type hinting just a bit neater and easier because we don't have to worry about importing the `typing` generics; we can use `dict[str, str]` directly.

I find it a lot easier to read code with type hints than without; I think we'll have fewer bugs the more we can rely on accurate type hints and linting tools. But, yeah; that's at least partly my opinion and instinct. There's probably data out there to be found if we want to bother to look.

Reply:
The challenge here is that we're using a Python dictionary as a generic "container", so the values could be strings, integers, or...other dictionaries (which are themselves string-string mappings)...so, that necessitates a somewhat more arcane description.  The parsing difficulty is partly that we're not used to it, and it's partly that the code's usage of this variable is simply that complex.

As for the value, well, that expression describes what we actually do.  I could use a cop-out like this:
```
    tms: Dict[str, Any] = {}
```
but, I feel like that provides less value.  (That is, I don't think the complicated type hint particularly hurts anything, so I appreciate the specificity.)  We could define a custom type for this, but I'm not convinced that that would improve the situation (especially since I don't think this type is used anywhere else).

Reply:
_[I didn't see the previous two posts when I posted my reply...the web page seems to be selective as to what it auto-updates.  🙂]_

I agree with Dave's points:  Python seems to be a superb language for prototyping, but it seems highly questionable that anyone would write anything "real" in it.  🤣  The fact that I can't tell whether the code will work until I run it (because we don't know what the types are until the code executes...and, possibly, not even then!) seems nigh on to silly.  Using type hints (followed by a code analyzer, such as Dave and I have in our respective IDEs) goes a long way toward alleviating those problems, and the more specific they are the better they work.  I'll echo Dave's comment that type hints, while not necessarily themselves making the declarations easier to read, do make the rest of the code much easier to understand (e.g., because you don't have to look at the **call sites** to know what types the parameters are supposed to have).  I think that, as we use them and get used to them, we'll start to value them more (and, at some point, we'll enable static analysis of our code, which will help even more).